src/server/gc_worker/gc_worker.rs

```diff
@@ -1219,21 +1221,25 @@ impl<E: Engine> GcWorker<E> {
             self.scheduler(),
             self.config_manager.clone(),
             self.feature_gate.clone(),
+            self.config_manager.value().thread_count,
         )
         .start()?;
         *handle = Some(new_handle);
         Ok(())
     }
 
     pub fn start(&mut self, store_id: u64) -> Result<()> {
-        let runner = GcRunner::new(
-            store_id,
-            self.engine.clone(),
-            self.flow_info_sender.take().unwrap(),
-            self.config_manager.0.clone().tracker("gc-woker".to_owned()),
-            self.config_manager.value().clone(),
-        );
-        self.worker.lock().unwrap().start(runner);
+        for _ in 0..self.config_manager.value().thread_count {
+            let runner = GcRunner::new(
+                store_id,
+                self.engine.clone(),
+                self.flow_info_sender.clone().unwrap(),
+                self.config_manager.0.clone().tracker("gc-woker".to_owned()),
+                self.config_manager.value().clone(),
+            );
+            self.worker.lock().unwrap().start(runner);
```

Comment:
Looking at LazyWorker (note: we're dispatching to LazyWorker here), what happens is that only one instance is started. In LazyWorker::start:

```diff
pub fn start<R: 'static + Runnable<Task = T>>(&mut self, runner: R) -> bool {
        if let Some(receiver) = self.receiver.take() {
            self.worker
                .start_impl(runner, receiver, self.metrics_pending_task_count.clone());
            return true;
        }
        false
    }
```

Here's a potential work around that uses a Mutex (as suggested by the rust book) to ensure that only one thread consumes from the queue at a time. I didn't spend much time on it, but perhaps this we could support this just for the GC use case:

```
+pub struct MultiThreadedWorker<T: Display + Send + 'static> {
+    scheduler: Scheduler<T>,
+    worker: Worker,
+    receiver: Arc<Mutex<UnboundedReceiver<Msg<T>>>>,
+    metrics_pending_task_count: IntGauge,
+}
+
+impl<T: Display + Send + 'static> MultiThreadedWorker<T> {
+
+    pub fn start<R: 'static + Runnable<Task = T>>(&mut self, runner: R) {
+        self.worker.start_multi_thread(runner, self.receiver.clone(), self.metrics_pending_task_count.clone());
+    }
+}
+
+
 impl<T: Display + Send + 'static> LazyWorker<T> {
     pub fn new<S: Into<String>>(name: S) -> LazyWorker<T> {
         let name = name.into();
@@ -470,6 +486,32 @@ impl Worker {
         });
     }
 
+    fn start_multi_thread<R: Runnable + 'static>(
+        &self,
+        runner: R,
+        rx: Arc<futures_util::lock::Mutex<UnboundedReceiver<Msg<R::Task>>>>,
+        metrics_pending_task_count: IntGauge,
+    )
+    where R::Task: Send,
+     {
+        let counter = self.counter.clone();
+        let _ = self.pool.spawn(async move {
+            let mut handle = RunnableWrapper { inner: runner };
+            loop {
+                let msg = rx.lock().await.next().await;
+                match msg {
+                    None => break,
+                    Some(Msg::Task(task)) => {
+                        handle.inner.run(task);
+                        counter.fetch_sub(1, Ordering::SeqCst);
+                        metrics_pending_task_count.dec();
+                    },
+                    Some(Msg::Timeout) => (),
+                }
+            };
+        });
+    }
+
```

Reply:
@afeinberg It would need to go through a lock for each task,  which is OK if the task is IO bounded. However as the general utility API we cannot assume the task is IO bounded. For these lightweight tasks,  going through a lock is not ideal.